File: mobile/packages/libghostty-terminal/android/src/main/java/expo/modules/libghostty/ExpoLibghosttyModule.kt
package expo.modules.libghostty

import android.util.Base64
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class ExpoLibghosttyModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoLibghostty")

    View(ExpoLibghosttyView::class) {
      Events("onInput", "onResize", "onBell", "onTitleChange", "onDirectoryChange", "onTap")

      // Base font size in dp (default 14); applied live, the grid reflows.
      Prop("fontSize") { view: ExpoLibghosttyView, size: Float ->
        view.setFontSize(size)
      }

      // Theme colors (ghostty config syntax); null clears to the defaults.
      Prop("theme") { view: ExpoLibghosttyView, theme: TerminalThemeRecord? ->
        view.setTheme(theme)
      }

      Prop("keyboardEnabled") { view: ExpoLibghosttyView, enabled: Boolean ->
        view.setKeyboardEnabled(enabled)
      }

      Prop("surfaceVisible") { view: ExpoLibghosttyView, visible: Boolean ->
        view.setSurfaceVisible(visible)
      }

      // PTY output bytes (base64) → terminal grid.
      AsyncFunction("write") { view: ExpoLibghosttyView, base64: String ->
        val data = try {
          Base64.decode(base64, Base64.DEFAULT)
        } catch (e: IllegalArgumentException) {
          throw InvalidBase64Exception(e)
        }
        view.write(data)
      }

      // PTY output as UTF-8 text → terminal grid.
      AsyncFunction("writeText") { view: ExpoLibghosttyView, text: String ->
        view.write(text.toByteArray(Charsets.UTF_8))
      }

      AsyncFunction("resetWithText") { view: ExpoLibghosttyView, text: String ->
        view.reset(text.toByteArray(Charsets.UTF_8))
      }

      AsyncFunction("pasteText") { view: ExpoLibghosttyView, text: String ->
        view.paste(text)
      }

      // PTY exited.
      AsyncFunction("finish") { view: ExpoLibghosttyView, exitCode: Int ->
        view.finish(exitCode)
      }

      OnViewDestroys { view: ExpoLibghosttyView ->
        view.destroy()
      }
    }
  }
}

internal class InvalidBase64Exception(cause: Throwable) :
  CodedException("expected base64-encoded terminal output", cause)

internal class TerminalThemeRecord : Record {
  @Field val background: String? = null
  @Field val foreground: String? = null
  @Field val cursorColor: String? = null
  @Field val selectionBackground: String? = null
  @Field val selectionForeground: String? = null

  /** Palette overrides by index (0-255) on top of ghostty's default palette. */
  @Field val palette: List<String?>? = null
}

File: mobile/packages/libghostty-terminal/ios/ExpoLibghosttyModule.swift
import ExpoModulesCore

public class ExpoLibghosttyModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoLibghostty")

    View(ExpoLibghosttyView.self) {
      Events("onInput", "onResize", "onBell", "onTitleChange", "onDirectoryChange", "onTap")

      // Base font size in points (default 14). Changing it after mount
      // rebuilds the surface and resets the grid.
      Prop("fontSize") { (view: ExpoLibghosttyView, size: Float?) in
        view.fontSize = size
      }

      // Theme colors (ghostty config syntax); nil clears to the defaults.
      // App-wide on iOS: applied through the shared controller's config.
      Prop("theme") { (view: ExpoLibghosttyView, theme: TerminalThemeRecord?) in
        view.applyTheme(theme)
      }

      Prop("keyboardEnabled") { (view: ExpoLibghosttyView, enabled: Bool) in
        view.keyboardEnabled = enabled
      }

      Prop("surfaceVisible") { (view: ExpoLibghosttyView, visible: Bool) in
        view.surfaceVisible = visible
      }

      // PTY output bytes (base64) → terminal grid.
      AsyncFunction("write") { (view: ExpoLibghosttyView, base64: String) in
        guard let data = Data(base64Encoded: base64) else {
          throw InvalidBase64Exception()
        }
        view.write(data)
      }.runOnQueue(.main)

      // PTY output as UTF-8 text → terminal grid.
      AsyncFunction("writeText") { (view: ExpoLibghosttyView, text: String) in
        view.write(text)
      }.runOnQueue(.main)

      AsyncFunction("resetWithText") { (view: ExpoLibghosttyView, text: String) in
        view.reset(with: text)
      }.runOnQueue(.main)

      AsyncFunction("pasteText") { (view: ExpoLibghosttyView, text: String) in
        view.paste(text)
      }.runOnQueue(.main)

      // PTY exited.
      AsyncFunction("finish") { (view: ExpoLibghosttyView, exitCode: UInt32) in
        view.finish(exitCode: exitCode)
      }.runOnQueue(.main)
    }
  }
}

internal final class InvalidBase64Exception: Exception {
  override var reason: String {
    "expected base64-encoded terminal output"
  }
}

internal struct TerminalThemeRecord: Record {
  @Field var background: String?
  @Field var foreground: String?
  @Field var cursorColor: String?
  @Field var selectionBackground: String?
  @Field var selectionForeground: String?

  /// Palette overrides by index (0-255) on top of ghostty's default palette.
  @Field var palette: [String?]?
}
